packager/certfetcher/certfetcher.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package certfetcher
+
+import (
+	"crypto"
+	"crypto/x509"
+	"strconv"
+
+	"github.com/WICG/webpackage/go/signedexchange"
+	"github.com/go-acme/lego/v3/certcrypto"
+	"github.com/go-acme/lego/v3/certificate"
+	"github.com/go-acme/lego/v3/challenge/http01"
+	"github.com/go-acme/lego/v3/lego"
+	"github.com/go-acme/lego/v3/registration"
+	"github.com/pkg/errors"
+)
+
+type CertFetcher struct {
+	// ACME Discovery URL
+	AcmeDiscoURL string
+	AcmeUser     MyUser
+	// Domains to validate
+	Domains    []string
+	legoClient *lego.Client
+}
+
+// You'll need a user or account type that implements acme.User
+type MyUser struct {
+	Email        string
+	Registration *registration.Resource
+	key          crypto.PrivateKey
+}
+
+func (u *MyUser) GetEmail() string {
+	return u.Email
+}
+func (u MyUser) GetRegistration() *registration.Resource {
+	return u.Registration
+}
+func (u *MyUser) GetPrivateKey() crypto.PrivateKey {
+	return u.key
+}
+
+// Initializes the cert fetcher with information it needs to fetch new certificates in the future.
+func NewFetcher(email string, privateKey crypto.PrivateKey, acmeDiscoURL string,
+	domains []string, acmeChallengePort int, shouldRegister bool) (*CertFetcher, error) {
+	myUser := MyUser{
+		Email: email,
+		key:   privateKey,
+	}
+	config := lego.NewConfig(&myUser)
+
+	config.CADirURL = acmeDiscoURL
+	config.Certificate.KeyType = certcrypto.RSA2048
+
+	// A client facilitates communication with the CA server.
+	client, err := lego.NewClient(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// We specify an http port of `acmeChallengePort`
```

Comment:
My 2c based on what I know so far is that we should lean towards the reverse proxy if we're picking exactly one.

The idea here is that this may be running on some remote server in a Google Cloud, Amazon, etc installation where shared filesystem may be difficult. The publisher already needs to set up reverse proxy rules for the Packager anyway, so adding one new rule should be fairly easy in any configuration.

Reply:
Good point, and I think agreed. If most people are using some CDN for their TLS termination, then they're probably using ACME DNS-01 challenge type, which won't interfere. This is probably the more common use-case.

If people are going full DIY, running their own Apache/nginx with [certbot](https://certbot.eff.org/) to manage their own TLS certs, then they're using HTTP-01 challenge type ([certbot doc](https://certbot.eff.org/docs/challenges.html)), so it would have to share the `/.well-known/acme-challenge/` URL path prefix. I'm not sure if there's a way to tell Apache/nginx to say "rev-proxy only if there isn't a file on disk".